pkg/util/client/client.go

```diff
@@ -31,23 +31,45 @@ func CreatePatch(before, after client.Object) (client.Patch, error) {
 	return client.RawPatch(patchBase.Type(), patchBytes), nil
 }
 
-func Patch(ctx context.Context, c client.Client, before, after client.Object) error {
-	patch, err := CreatePatch(before, after)
+// Patch applies the merge patch of client.Object with SSA.
+// If strict is true, the resourceVersion will be part of the patch, make this call fail if
+// client.Object was changed.
+func Patch(ctx context.Context, c client.Client, obj client.Object, strict bool, update func() (bool, error)) error {
+	objOriginal := obj.DeepCopyObject().(client.Object)
+	if strict {
+		objOriginal.SetResourceVersion("")
+	}
+	updated, err := update()
+	if err != nil || !updated {
+		return err
+	}
+	patch, err := CreatePatch(objOriginal, obj)
 	if err != nil {
 		return err
 	}
-	if err = c.Patch(ctx, before, patch); err != nil {
+	if err = c.Patch(ctx, objOriginal, patch); err != nil {
 		return err
 	}
 	return nil
 }
 
-func PatchStatus(ctx context.Context, c client.Client, before, after client.Object) error {
-	patch, err := CreatePatch(before, after)
+// PatchStatus applies the merge patch of client.Object status with SSA.
+// If strict is true, the resourceVersion will be part of the patch, make this call fail if
+// client.Object was changed.
+func PatchStatus(ctx context.Context, c client.Client, obj client.Object, strict bool, update func() (bool, error)) error {
```

Comment:
Are we ever using strict=false?

Reply:
Almost on all cases we are using `strict=true`, except

https://github.com/kubernetes-sigs/kueue/blob/4f3689bc803da641f761f978142b92dbb8b92ff7/pkg/controller/jobs/job/job_controller.go#L166

After research an [error](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/2553/pull-kueue-test-integration-main/1810412244677169152) I found that `Get` sometimes return old version of `Job`.

```go
	if !j.IsSuspended() {
		if err := clientutil.Patch(ctx, c, object, func() (bool, error) {
			j.Suspend()
			return true, nil
		}); err != nil {
			return false, fmt.Errorf("suspend: %w", err)
		}
		if err := c.Get(ctx, client.ObjectKeyFromObject(object), object); err != nil {
			return false, fmt.Errorf("get after suspend: %w", err)
		}
		stoppedNow = true
	}
```

And in the next updates we will have a conflict error.

So in this case we can use `strict=false` to avoid this error.

Reply:
This is worrisome, I see that `Get` could occasionally return the old object, but in this case it should succeed on retry. Wasn't this retried in the test, or was retried, but fail consistently every time?

Reply:
Yes, almost everytime I got this error.

Reply:
The biggest problem that we can't execute `job.Stop()` second time if we will have an error after suspend.

Reply:
https://github.com/kubernetes-sigs/kueue/blob/4f3689bc803da641f761f978142b92dbb8b92ff7/pkg/controller/jobframework/reconciler.go#L602-L623

Reply:
> The biggest problem that we can't execute job.Stop() second time if we will have an error after suspend.

Do you mean that re-try will not be attempted - skiped?

Does it mean we also had an issue before - that if Update to reset start time fails, then we would also skip retry, and never fully reset the start time?

Reply:
Looks like yes :(

Reply:
I think to retry we could probably just remove this check: `!job.IsSuspended()` and leave `if match == nil`. Can you check if this helps?

If this helps, then this is something we should also do for Updates. What is a bit worrisome in case of patches, is that the get often gets the old version, and so results in common errors.

Reply:
I don't believe we had a bug in this logic before. We used to have a lot of bug reports on this topic, but we haven't had one in a couple of months.
Let's not change the behavior, just swap every Update with a Patch with resourceVersion, and the outcome should be exactly the same.

Reply:
But we have an [issue](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/2553/pull-kueue-test-integration-main/1810412244677169152) that approve it.

Reply:
@mimowo I agree with you to open another PR. But without fix this issue I can't use `strict=true` on `job.Stop()` :(

Reply:
Created an issue #2566

Reply:
I see, thanks! We should be able to fix that independently.